src/basic/namespace-util.c

```diff
@@ -210,14 +213,40 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
 
         /* Forks off a process in a new userns, configures the specified uidmap/gidmap, acquires an fd to it,
          * and then kills the process again. This way we have a userns fd that is not bound to any
-         * process. We can use that for file system mounts and similar. */
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure the mapping is done after setuid(). */
 
-        r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
+        if (pipe2(pipe_fd, 0) < 0)
+                return log_error_errno(errno, "Failed to open bidirectional pipe: %m");
+
+        /* If we have to set a different owner we have to setuid before unsharing the new namespace, otherwise
+         * the kernel will refuse the operation. */
+        if (owner == UID_INVALID)
+                flags |= FORK_NEW_USERNS;
+
+        r = safe_fork_full("(sd-mkuserns)", (int[]) { pipe_fd[1] }, 1, flags, &pid);
         if (r < 0)
                 return r;
-        if (r == 0)
-                /* Child. We do nothing here, just freeze until somebody kills us. */
+        if (r == 0) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner != UID_INVALID) {
+                        (void) setuid(owner);
+                        (void) unshare(CLONE_NEWUSER);
```

Comment:
Why do you ignore return values here? Both can legitimately fail. I would check for failures in the child. Also, the freeze thing seems strange and is that even guaranteed to be correctly ordered such that the parent is guaranteed to try to write that mapping after the userns is unshared?

I would redo this and align this with how this is more commonly done:
* fork child and use socketpair()
* blocking read in parent
* let child to setuid and create userns. then write to socket to let parent know that we're ready for it to write mapping
* read from socket succeeds and child told us that we can write the mapping
* write the idmapping for the child and open the namespace file descriptor
* inform child that we're done writing the mapping and close socket
* wait for child to exit

This is way cleaner and guarantees correct ordering under all circumstances.

Reply:
the freeze stuff is pre-existing, and closing the pipe is the same as writing to a socket and it's how it's done elsewhere

Reply:
> Why do you ignore return values here? Both can legitimately fail. I would check for failures in the child

Follow-up operations will fail if those don't work, so there's no need for extra complications with transferring error codes over

Reply:
The resulting namespace cannot be used by the caller, joining it will fail with permission denied, I'm pretty sure